internal/channel/channel.go

```diff
@@ -20,109 +19,149 @@ type Channel struct {
 	Config string `db:"config" json:"-"` // excluded from JSON config dump as this may contain sensitive information
 
 	Logger *zap.SugaredLogger
-	cmd    *exec.Cmd
-	rpc    *rpc.RPC
-	mu     sync.Mutex
-}
 
-func (c *Channel) Start(pluginDir string) error {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+	newConfigCh chan struct{}
+	pluginCh    chan *Plugin
 
-	if c.cmd != nil && c.rpc.Err() == nil {
-		return nil
-	}
+	pluginCtx       context.Context
+	pluginCtxCancel func()
+}
 
-	if matched, _ := regexp.MatchString("^[a-zA-Z0-9]*$", c.Type); !matched {
-		return fmt.Errorf("channel type must only contain a-zA-Z0-9, %q given", c.Type)
-	}
+// Start initializes the channel and starts the plugin in the background
+func (c *Channel) Start(logger *zap.SugaredLogger, ctx context.Context) {
+	c.Logger = logger
+	c.newConfigCh = make(chan struct{})
+	c.pluginCh = make(chan *Plugin)
+	c.pluginCtx, c.pluginCtxCancel = context.WithCancel(ctx)
 
-	path := filepath.Join(pluginDir, c.Type)
+	go c.runPlugin()
+}
 
-	cmd := exec.Command(path)
+// initPlugin returns a new Plugin or nil if an error occurred during initialization
+func (c *Channel) initPlugin() *Plugin {
+	c.Logger.Debug("Initializing channel plugin")
 
-	writer, err := cmd.StdinPipe()
+	p, err := NewPlugin(c.Type, c.Logger)
 	if err != nil {
-		return fmt.Errorf("failed to create stdin pipe: %w", err)
+		c.Logger.Errorw("Failed to initialize channel plugin", zap.Error(err))
+		return nil
 	}
 
-	reader, err := cmd.StdoutPipe()
-	if err != nil {
-		return fmt.Errorf("failed to create stdout pipe: %w", err)
+	if err := p.SetConfig(c.Config); err != nil {
+		c.Logger.Errorw("Failed to set channel plugin config, terminating the plugin", zap.Error(err))
+		p.Stop()
+		return nil
 	}
 
-	errPipe, err := cmd.StderrPipe()
-	if err != nil {
-		return fmt.Errorf("failed to create stderr pipe: %w", err)
-	}
+	return p
+}
 
-	go forwardLogs(errPipe, c.Logger)
+// runPlugin is called as go routine to initialize and maintain the plugin by receiving signals on given chan(s)
+func (c *Channel) runPlugin() {
+	var currentlyRunningPlugin *Plugin
+	for {
+		if currentlyRunningPlugin == nil {
+			currentlyRunningPlugin = c.initPlugin()
+		}
 
-	if err = cmd.Start(); err != nil {
-		return fmt.Errorf("failed to start cmd: %w", err)
-	}
-	c.Logger.Debug("Cmd started successfully")
+		var rpcDone <-chan struct{}
+		if currentlyRunningPlugin != nil {
+			rpcDone = currentlyRunningPlugin.rpc.Done()
+		}
```

Comment:
Moving this info a function should make this loop cleaner (that function can then just be called in the select below, no manually define a variable then). Maybe also just as a closure like the helper suggested in my next comment.